Source/DafnyCore/Dafny.atg

```diff
@@ -3202,6 +3203,52 @@ Forall = "forall".
 Exists = "exists".
 QSep = "::".
 
+Expression<out Expression e, bool allowLemma, bool allowLambda, bool allowBitwiseOps = true>
+=
+  NotDecreasesToExpression<out e, allowLemma, allowLambda, allowBitwiseOps>
+  .
+
+MaybeDecreasesToExpression<out Expression e, IToken lp>
+= (. List<ActualBinding> args = new();
+     List<Expression> newExprs = new();
+     Expression tmp = null;
+     IToken x = null;
+     bool allowNoChange = false;
+  .)
+  TupleArgs<args>
+  [ ( "decreases"                                                               (. allowNoChange = false; .)
+    | "nonincreases"                                                            (. allowNoChange = true; .)
+    ) (. x = t; .)
+    ident
```

Comment:
How come you can't write "to" here? 

Also, have you considered not having `to` at all? Having two keywords that always follow each other with a space in between is something that I do not think we have anywhere else it the language. It looks like we're trying to mimic natural language, which I don't think is our goal.

If you want to keep the `to`, I would remove the space, so you get `decreasesto`, similar to `nonincreases`, but I would prefer `decreases`.

To be honest, actually I would prefer introducing built-in functions `decreasesTo` and `decreasesOrEqualTo`. Having a keyword as an infix operator feels off for the language. Do we do that anywhere else? Also, for something that occurs so little, I do not see a reason to have special syntax for it.

Also, your implementation always requires `decreases` to be surrounded by parenthesis. Is that what we settled on? It looks unnecessary: `assert (true decreases to false);`. If you're requiring parenthesis anyways, then why not make it look like a function call: `assert decreasesTo(true, false);` ?

Reply:
> How come you can't write "to" here?

We could only do that if we make it a keyword, which I'd rather not do. The parsing for ranges in things like `for var x := 0 to 10` works like this, too.

> Also, have you considered not having `to` at all? Having two keywords that always follow each other with a space in between is something that I do not think we have anywhere else it the language. It looks like we're trying to mimic natural language, which I don't think is our goal.

To me, it seems much harder to understand without the "to".

> If you want to keep the `to`, I would remove the space, so you get `decreasesto`, similar to `nonincreases`, but I would prefer `decreases`.

I originally implemented it that way, for simplicity, but we had agreed on having the space in earlier discussions, so I migrated to that syntax.

> To be honest, actually I would prefer introducing built-in functions `decreasesTo` and `decreasesOrEqualTo`. Having a keyword as an infix operator feels off for the language. Do we do that anywhere else? Also, for something that occurs so little, I do not see a reason to have special syntax for it.

There needs to be some degree of special syntax because there are two arbitrarily long tuples of expressions on either side. For ease of parsing, my initial experiments used `decreases[a, b ; a', b']`, but that feels very clunky to me.

> Also, your implementation always requires `decreases` to be surrounded by parenthesis. Is that what we settled on? It looks unnecessary: `assert (true decreases to false);`. If you're requiring parenthesis anyways, then why not make it look like a function call: `assert decreasesTo(true, false);` ?

It's close to necessary because of the fact that you can have an arbitrary number of commas on either side of `decreases to`. It would be possible to allow it as the top level expression without parentheses in things like `assert`, `assume`, `requires`, and `ensures`, though I believe it would require substantial duplication in the grammar to do that. I was on the fence about whether to implement that or not. Doing so would be entirely backward-compatible with what's in this PR, though, so I'd prefer to make smaller changes per PR and do it as a separate one (if at all).

Reply:
> There needs to be some degree of special syntax because there are two arbitrarily long tuples of expressions on either side. For ease of parsing, my initial experiments used decreases[a, b ; a', b'], but that feels very clunky to me.

I see, you need a divider for the two lists.

> It's close to necessary because of the fact...

I think at this point I'm in favor of using tuple syntax on both sides if there is more than one item to be compared, so `(a,b) decreases to (c,d)`, instead of `(a, b decreases to c, d)` (which reads like a 3-tuple with `b decreases to c` in the middle), and `a decreases to b` instead of `(a decreases to b)`. Part of the reason I prefer this is because it's most similar to `(a,b) > (c,d)`. I vaguely recall we decided on `(a, b decreases to c, d)` in the syntax discussion meeting, and that I was also in favor of this, but looking at it now I'm not in favor.

>> If you want to keep the to, I would remove the space, so you get decreasesto, similar to nonincreases, but I would prefer decreases.

> I originally implemented it that way, for simplicity, but we had agreed on having the space in earlier discussions, so I migrated to that syntax.

What's your opinion? Would you be willing to still do a tiny poll amongst the team? I recall we had group discussion on the syntax already, but it might be easier to weigh small choices such as this one separately.

Reply:
> What's your opinion? Would you be willing to still do a tiny poll amongst the team? I recall we had group discussion on the syntax already, but it might be easier to weigh small choices such as this one separately.

Oh, yeah, I'm definitely up for changing it if the general consensus is toward a different syntax. I implemented this as the closest feasible thing to what we'd agreed on.

The code we're discussing here has already been merged, but changes are very unlikely to break anything if we do them relatively soon.

Reply:
I get the issue of the 3-tuple here, but at the same time (a, b) would feel like a tuple which has different decreases to rules. At this point, I prefer the syntax that looks like the 3-tuple here, although in practice I'll surely format it as
```(a, b
    decreases to
    c, d)
```
or
```(a, b
    decreases to c, d)
```

the same way we usually do for other infixes operators.

Reply:
Yeah, exactly. Said a different way, and using the syntax I implemented, the semantics of `(a, b decreases to c, d)` are different than the semantics of `((a, b) decreases to (c, d))`. So how would we write the second one if we represented both sides as tuples? Would we do `(((a, b)) decreases to ((c, d)))`? That seems weird, because normally parentheses are intended to modify only syntax, not semantics.

Reply:
> I get the issue of the 3-tuple here, but at the same time `(a, b)` would feel like a tuple which has different decreases to rules.

Indeed `(a, b)` looks like a tuple. `decreases to` would then be an operator that takes two tuples as arguments, applies Dafny's `<` operator on the tuple elements, and then does a lexicographic comparison on those results.